data/transactions/logic/eval.go

```diff
@@ -1350,23 +1397,21 @@ func (cx *EvalContext) step() error {
 			return fmt.Errorf("%3d %s returned 0 cost", cx.pc, spec.Name)
 		}
 	}
-	cx.cost += opcost
-	if cx.PooledApplicationBudget != nil {
-		*cx.PooledApplicationBudget -= opcost
-	}
 
-	if cx.remainingBudget() < 0 {
-		// We're not going to execute the instruction, so give the cost back.
-		// This only matters if this is an inner ClearState - the caller should
-		// not be over debited. (Normally, failure causes total txtree failure.)
-		cx.cost -= opcost
-		if cx.PooledApplicationBudget != nil {
-			*cx.PooledApplicationBudget += opcost
-		}
+	if opcost > cx.remainingBudget() {
 		return fmt.Errorf("pc=%3d dynamic cost budget exceeded, executing %s: local program cost was %d",
 			cx.pc, spec.Name, cx.cost)
 	}
 
+	cx.cost += opcost
+	// Only one of these pooled budgets will be non-nil, perhaps we could collapse to one variable.
```

Comment:
can't you check cx.ep.runmode?

Reply:
We could, do you like that better?  That would be a switch, and then we'd have each of these tests in the respective arm of the switch anyway. (we need to test non-nil no matter what, since pre-pooling, these budgets are also nil).

So by wrapping everything in switch arms, we strictly add code. Maybe it's clearer though?

Reply:
We ain't got no fancy ternary operator round these parts.

I do like avoiding the second check if the first is true.  How about:
```
switch {
	case cx.PooledApplicationBudget != nil:
		*cx.PooledApplicationBudget -= opcost
	case cx.PooledLogicSigBudget != nil {
		*cx.PooledLogicSigBudget -= opcost
}
```
(Good point to do apps first, as we care more about perf there.)